File: eval-platform/python_client/player/services/strategy_service.py
from player.models import KitClasses, RoundRequest

class StrategyService:
    def __init__(self, aircraft_map, logger):
        self.aircraft_map = aircraft_map
        self.logger = logger
        self.pending_loads = []

    def analyze_events(self, events):
        for event in events:
            if event.event_type == "CHECKED_IN":
                self._calculate_load(event)

    def apply_decisions(self, round_request):
        for load in self.pending_loads:
            round_request.flight_loads.append(load)
        
        # Clear local memory
        self.pending_loads.clear()

    def _calculate_load(self, event):
        ac_type = self.aircraft_map.get(event.aircraft_type)

        # LOGIC: Survival Mode (Load exactly what is needed)
        load_e = min(event.passengers.economy, ac_type.economy_capacity)
        load_b = min(event.passengers.business, ac_type.business_capacity)
        load_f = min(event.passengers.first, ac_type.first_capacity)
        load_p = min(event.passengers.premiumEconomy, ac_type.premium_capacity)

        kits = KitClasses(load_f, load_b, load_p, load_e)
        
        self.logger.info(f"-> [ORDERING] Flight {event.flight_number}: {kits}")
        
        # Store as the dictionary structure expected by RoundRequest
        self.pending_loads.append({
            "flightId": event.flight_id,
            "loadedKits": kits.to_dict()
        })

File: eval-platform/src/main/java/com/sap/hackaton2025/player/service/StrategyService.java
package com.sap.hackaton2025.player.service;

import com.sap.hackaton2025.player.model.*;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

public class StrategyService {
    
    private final Map<String, AircraftType> aircraftMap;
    private final List<RoundRequest.FlightLoadDto> pendingLoads = new ArrayList<>();

    private final LogService logger; 

    public StrategyService(Map<String, AircraftType> aircraftMap, LogService logger) {
        this.aircraftMap = aircraftMap;
        this.logger = logger;
    }

    public void analyzeEvents(List<FlightEvent> events) {
        for (FlightEvent event : events) {
            if ("CHECKED_IN".equals(event.eventType)) {
                calculateLoadForFlight(event);
            }
        }
    }
    
    public void applyDecisionsToRequest(RoundRequest request) {
        request.flightLoads.addAll(pendingLoads);        
        pendingLoads.clear();
    }

    private void calculateLoadForFlight(FlightEvent event) {
        AircraftType type = aircraftMap.get(event.aircraftType);
        if (type == null) return;

        // Load exactly the number of passengers checked in, capped by aircraft capacity
        int loadE = Math.min(event.passengers.economy, type.economyKitsCapacity);
        int loadB = Math.min(event.passengers.business, type.businessKitsCapacity);
        int loadF = Math.min(event.passengers.firstClass, type.firstClassKitsCapacity);
        int loadP = Math.min(event.passengers.premiumEconomy, type.premiumEconomyKitsCapacity);

        // Create a ProcessBuilder to call a python script that will call a print

        ProcessBuilder pb = new ProcessBuilder("python3", "scripts/load.py",
                event.flightNumber,
                String.valueOf(loadF),
                String.valueOf(loadB),
                String.valueOf(loadP),
                String.valueOf(loadE)
        );

        pb.inheritIO();

        try {
            Process process = pb.start();
            process.waitFor();
        } catch (Exception e) {
            logger.info("Eroare la apelarea scriptului de logare a incarcarii: " + e.getMessage());
        }

        KitClasses load = new KitClasses(loadF, loadB, loadP, loadE);
        
        logger.info("   -> [CHECK-IN] Zbor " + event.flightNumber + " (" + event.flightId + "): " + load.toString());
        pendingLoads.add(new RoundRequest.FlightLoadDto(event.flightId, load));
    }
}